### PR TITLE
Visual Editor: Keep track of the focused block

### LINKED ISSUE #437: Cannot type in empty image caption

### PR DESCRIPTION
**Pull request by youknowriad**
references #420 

In this PR, I'm adding the `focus` state to keep track of the focused block and any extra config that let us focus the right element. For now, I'm storing the focused block and the focused editable inside the block. To make undo/redo work correctly, we'll need to add the position inside the focused editable.

closes #437

### CODE REVIEW

File: blocks/library/image/index.js [hunk 1]
```diff
@@ -19,21 +19,26 @@ registerBlock( 'core/image', {
 		caption: html( 'figcaption' )
 	},
 
-	edit( { attributes, isSelected, setAttributes } ) {
+	edit( { attributes, setAttributes, focus, updateFocus } ) {
 		const { url, alt, caption } = attributes;
+		const focusCaption = () => updateFocus( { editable: 'caption' } );
 
+		/* eslint-disable */
```
Comment (youknowriad) [inline blocks/library/image/index.js]: the `onClick` adds to many accessibility linting errors :(

Comment (jasmussen) [inline blocks/library/image/index.js]: Not sure of the context, but is it worth adding a note about this to a ticket like https://github.com/WordPress/gutenberg/issues/392?

File: editor/state.js [hunk 2]
```diff
@@ -131,6 +131,25 @@ export function hoveredBlock( state = null, action ) {
 }
 
 /**
+ * Reducer returning the focused block state.
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Dispatched action
+ * @return {Object}        Updated state
+ */
+export function focus( state = {}, action ) {
```
Comment (aduth) [inline editor/state.js]: Should this be included in undo history?

File: blocks/library/image/index.js (hunk 1 above)
Comment (aduth) [inline blocks/library/image/index.js]: Also, we should be specifically disable only the lint rule we want to target, in case there are other lint issues in the region.

I like the idea of commenting on any disable we apply, since it holds us accountable to understanding why lint errors occur and how they're not applicable.

File: blocks/library/image/index.js [hunk 3]
```diff
@@ -19,21 +19,26 @@ registerBlock( 'core/image', {
...
 			<figure>
-				<img src={ url } alt={ alt } />
-				{ caption || isSelected ? (
+				<img src={ url } alt={ alt } onClick={ focusCaption } />
+				{ caption || !! focus ? (
 					<Editable
 						tagName="figcaption"
 						placeholder={ wp.i18n.__( 'Write caption…' ) }
 						value={ caption }
+						focus={ focus && focus.editable === 'caption' ? focus : null }
```
Comment (aduth) [inline blocks/library/image/index.js]: Do we even need the config here since we only have one focusable element? Wondering if `focusCaption` could be implemented just as:

```
const focusCaption = () => setFocus();
```

And we'd infer caption should be focused merely by a truthy incoming `focus` value?

File: blocks/library/quote/index.js [hunk 4]
```diff
@@ -7,7 +7,7 @@ import Editable from 'components/editable';
 
 const { parse, html, query } = hpq;
 
-const fromValueToParagraphs = ( value ) => value.map( ( paragraph ) => `<p>${ paragraph }</p>` ).join( '' );
+const fromValueToParagraphs = ( value ) => value ? value.map( ( paragraph ) => `<p>${ paragraph }</p>` ).join( '' ) : '';
```
Comment (aduth) [inline blocks/library/quote/index.js]: Is this fixing a bug? Do we really want to support an undefined `value`? Do we need some concept of default attributes, whether that's an explicit property of a block or even just how we destructure `attributes` in `edit` and `save`?

**Review by aduth — COMMENTED** [review]
This seems reasonable to me. As with anything we add, we should be conscious of the developer experience impact. I think @iseulde raises a good point at https://github.com/WordPress/gutenberg/issues/420#issuecomment-294480454 . At a glance, "focused" and "selected" seem synonymous, so how can we clarify or consolidate the two? With regards to new props being passed to `edit`, are there options to explore with [React's hidden context feature](https://facebook.github.io/react/docs/context.html) to help avoid the need for the block implementer to consider managing focus themselves?

File: editor/state.js (hunk 2 above)
Comment (youknowriad) [inline editor/state.js]: Probably 👍

File: blocks/library/image/index.js (hunk 1 above)
Comment (youknowriad) [inline blocks/library/image/index.js]: Yep that was my intent, but this is breaking three rules :). I'll update.

File: blocks/library/image/index.js (hunk 3 above)
Comment (youknowriad) [inline blocks/library/image/index.js]: I hesitated and you're probably right here.

File: blocks/library/quote/index.js (hunk 4 above)
Comment (youknowriad) [inline blocks/library/quote/index.js]: Yes this fixes the bug of inserting an empty quote block (using the inserter) since we're assuming the attributes are all nullable.

File: editor/state.js (hunk 2 above)
Comment (youknowriad) [inline editor/state.js]: Tried this and it has several implications, we should ignore consecutive focuses etc... I think this should be added but in its own PR.

**Comment by youknowriad** [comment]
@aduth I'm not sure how can we prevent the block author to handle the focus

**Comment by youknowriad** [comment]
>  I think @iseulde raises a good point at #420 (comment) . At a glance, "focused" and "selected" seem synonymous, so how can we clarify or consolidate the two? 

I agree, and I'm really bad with naming things. Open to any suggestion.

**Comment by aduth** [comment]
>@aduth I'm not sure how can we prevent the block author to handle the focus

Well, at first I'd thought to use [React's context](https://facebook.github.io/react/docs/context.html) to have the visual editor define child context values that the `Editable` component could directly access. This is how `react-redux` gets around the need to pass store as a prop down through intermediary components. But thinking about `react-redux` got me to consider whether it could interact with the store directly? The obvious problem is that `Editable` is supposed to not be aware of the editor, being that it's a general purpose component in the `blocks` directory. Per #408, I think we'll need to find a better solution that lets us treat components as editor-aware while still exposing them for consumption to external plugins.

**Comment by aduth** [comment]
>I agree, and I'm really bad with naming things. Open to any suggestion.

I'm not totally sure it's entirely a naming problem, as I don't really see much of a conceptual difference between our idea of selected and focused, aside from the fact that "selected" accounts for typing status to fade out the border and controls. And if that's the case (correct me if it's not), could this be better reflected by a standalone `isTyping` state value updated in response to a global keyboard event?

**Comment by youknowriad** [comment]
@aduth @iseulde Good point.

In e3ce9f6 I merged the "selected" and the "focused" concepts to avoid confusion.

@aduth I'm not certain we should hide the "focus" management from the block author. I think the block author should have some control on keyboard navigation and stuff like that (think moving the cursor from the content to the citation in the quote block). I think we should merge this as is and explore enhancements.

File: editor/modes/visual-editor/block.js [hunk 5]
```diff
@@ -58,13 +57,13 @@ function VisualEditorBlock( props ) {
 			tabIndex="0"
 			onFocus={ onSelect }
 			onBlur={ maybeDeselect }
-			onKeyDown={ onDeselect }
-			onMouseEnter={ onMouseEnter }
+			onKeyDown={ onStartTyping }
+			onMouseMove={ onMouseMove }
```
Comment (aduth) [inline editor/modes/visual-editor/block.js]: `onMouseMove` is an incredibly noisy event (watch Redux DevTools while moving). Is the idea here that we want the block to regain its hover effects when moving the mouse while typing? Maybe we can at least add a condition to only trigger the callback if `isTyping` is true?

File: editor/modes/visual-editor/block.js [hunk 6]
```diff
@@ -73,9 +72,11 @@ function VisualEditorBlock( props ) {
 					} ) ) } />
 			) : null }
 			<BlockEdit
-				isSelected={ isSelected }
+				focus={ focus }
 				attributes={ block.attributes }
-				setAttributes={ setAttributes } />
+				setAttributes={ setAttributes }
+				updateFocus={ onFocus }
```
Comment (aduth) [inline editor/modes/visual-editor/block.js]: Even just for consistency with `setAttributes`, do you think naming this `setFocus` is still accurate / any better?

File: editor/modes/visual-editor/block.js [hunk 7]
```diff
@@ -122,6 +131,13 @@ export default connect(
 				hovered: false,
 				uid: ownProps.uid
 			} );
+		},
+		onFocus( config = {} ) {
```
Comment (aduth) [inline editor/modes/visual-editor/block.js]: Should we be assigning the default config into state memory or would it be more appropriate to apply the default in `mapStateToProps`?

File: editor/state.js [hunk 8]
```diff
@@ -92,17 +92,49 @@ export const blocks = combineUndoableReducers( {
...
+
+		case 'UPDATE_FOCUS':
+			return {
+				uid: action.uid,
+				typing: state.uid === action.uid ? state.typing : false,
+				focus: action.config
+			};
+
+		case 'START_TYPING':
+			if ( action.uid !== state.uid ) {
```
Comment (aduth) [inline editor/state.js]: Does typing consideration need to be per-block? Seems like a global effect.

**Comment by aduth** [comment]
Needs a rebase.

File: editor/state.js (hunk 8 above)
Comment (youknowriad) [inline editor/state.js]: Like implemented for now "START_TYPING" is per-block. It's triggered when typing inside a specific block, we can drop the `uid`. I feel it should stay as is unless we catch the start typing in a higher level maybe?

File: editor/modes/visual-editor/block.js (hunk 5 above)
Comment (youknowriad) [inline editor/modes/visual-editor/block.js]: Exactly! yep good ideal to limit it when we're typing.

File: editor/modes/visual-editor/block.js (hunk 7 above)
Comment (youknowriad) [inline editor/modes/visual-editor/block.js]: Right, handling the default value in the state is safer

**Comment by iseulde** [comment]
If it helps, in previous prototypes I had also a separate key for showing and hiding UI in the Redux store. `SHOW_UI` on `mousedown touchstart` and `HIDE_UI` on `keydown blur`. There are state checks before dispatching.

**Comment by youknowriad** [comment]
Feedback addressed, should we merge?

File: blocks/components/editable/index.js [hunk 9]
```diff
@@ -115,7 +133,16 @@ export default class Editable extends wp.element.Component {
 	componentDidUpdate( prevProps ) {
 		if ( this.props.tagName !== prevProps.tagName ) {
 			this.initialize();
-		} else if ( this.props.value !== prevProps.value ) {
+		}
+
+		if ( this.props.focus !== prevProps.focus && this.props.focus ) {
```
Comment (aduth) [inline blocks/components/editable/index.js]: Because of the strict comparison of objects and how the reducer is creating next state, this condition passes much more frequently than we'd probably expect. For example I see upon adding a `console.log` to the inner block that it gets called twice every time I select a block.

Comment (nylen) [inline blocks/components/editable/index.js]: Maybe `if ( this.props.focus && ! prevProps.focus )` would be better?

**Comment by nylen** [comment]
What is `focus`?  How can we make this more clear?

Why is it an object rather than a primitive such as a boolean (or a string to indicate that a specific field is focused)?  If it's possible to make it a primitive, this would also solve https://github.com/WordPress/gutenberg/pull/438#discussion_r112246377 at the same time.

**Comment by nylen** [comment]
Also, from a usability perspective, this seems to be working extremely well and addresses a bunch of issues I was having previously, so huge :+1: there

File: blocks/components/editable/index.js (hunk 9 above)
Comment (youknowriad) [inline blocks/components/editable/index.js]: I think to avoid these calls, we could check for a falsy `this.props.focus` before calling the `onFocus` prop for now. Once we introduce the focus position (probably a following PR) we should check if it's the same position before calling the callback.

**Comment by youknowriad** [comment]
@nylen `this.props.focus` is a nullable object. If it's falsy, this means the current block does not have the focus. If it's not, its shape is defined partially in this PR. and the final shape could be `{ editable: 'citation', offset: 10 }` (maybe tinymce's bookmark instead of offset, not sure yet). 

The `editable` is optional. It's intended for blocks with multiple editables and the offset is the offset inside this editable.

**Comment by nylen** [comment]
OK, thanks for explaining.  How can we document this in the code?  Should it be two separate properties instead?

Probably the solution to https://github.com/WordPress/gutenberg/pull/438#discussion_r112246377 then is to compare each of the primitive properties inside of `focus`.

File: blocks/components/editable/index.js (hunk 9 above)
Comment (youknowriad) [inline blocks/components/editable/index.js]: @nylen proposition is valid for now because we don't store any offset yet. I updated in 7488f23.

**Comment by youknowriad** [comment]
Feedback addressed, anyone to give a last 👀 before merging?

**Review by nylen — APPROVED** [review]
As Riad pointed out in Slack, changing `focus` to primitive properties would complicate the code.  We can document afterwards along with the rest of the `edit` parameters.  I'll make a separate issue for that.
